### PR TITLE
Fix stack overflow detection on FreeBSD

### PR DESCRIPTION
**Pull request by asomers**

### CODE REVIEW

File: src/libstd/sys/unix/thread.rs [hunk 1]
```diff
@@ -199,6 +201,22 @@ pub mod guard {
...
+        if libc::pthread_attr_get_np(libc::pthread_self(), &mut attr) == 0 {
+            let mut stackaddr = ::ptr::null_mut();
+            let mut stacksize = 0;
+            assert_eq!(libc::pthread_attr_getstack(&attr, &mut stackaddr,
+                                                   &mut stacksize), 0);
+            ret = Some(stackaddr);
+        }
+        assert_eq!(libc::pthread_attr_destroy(&mut attr), 0);
+        ret
+    }
```
Comment (alexcrichton) [inline src/libstd/sys/unix/thread.rs]: This looks pretty similar to `get_stack_start` with Linux/Android/NetBSD below, perhaps the two could share code? If the only difference is the name of the function `pthread_getattr_np`, then some renaming-on import should suffice I think.

File: src/libstd/sys/unix/thread.rs [hunk 2]
```diff
@@ -282,6 +304,27 @@ pub mod guard {
         })
     }
 
+    #[cfg(target_os = "freebsd")]
+    pub unsafe fn current() -> Option<usize> {
```
Comment (alexcrichton) [inline src/libstd/sys/unix/thread.rs]: Similar comment here as the one above.

File: src/libstd/sys/unix/thread.rs [hunk 3]
```diff
@@ -282,12 +293,18 @@ pub mod guard {
...
         let mut ret = None;
         let mut attr: libc::pthread_attr_t = ::mem::zeroed();
         assert_eq!(libc::pthread_attr_init(&mut attr), 0);
-        if libc::pthread_getattr_np(libc::pthread_self(), &mut attr) == 0 {
+        #[cfg(target_os = "freebsd")]
+            let e = libc::pthread_attr_get_np(libc::pthread_self(), &mut attr);
+        #[cfg(not(target_os = "freebsd"))]
+            let e = libc::pthread_getattr_np(libc::pthread_self(), &mut attr);
+        if e == 0 {
+        //if libc::pthread_getattr_np(libc::pthread_self(), &mut attr) == 0 {
```
Comment (sfackler) [inline src/libstd/sys/unix/thread.rs]: Dead code.

**Comment by asomers** [comment]
@alexcrichton does it look good to you now?

**Comment by alexcrichton** [comment]
@bors: r+ 78ea972b9ff4a2c3eaa0328f6fa8e7038e8ca9d1

Thanks! Looks good to me

**Pull request merged**